internal/engine/planner/planner.go

```diff
@@ -79,6 +81,43 @@ func CombinePlans(principalPolicyPlan, resourcePolicyPlan *PolicyPlanResult) *Po
 	}
 }
 
+func mergePlans(acc, current *PolicyPlanResult) *PolicyPlanResult {
+	if acc == nil {
+		return current
+	}
+	scopePermissions := current.ScopePermissions
+	allowFilter := current.AllowFilter
+	if current.AllowEmpty() {
+		scopePermissions = acc.ScopePermissions
+		allowFilter = acc.AllowFilter
+	} else if !acc.AllowEmpty() {
+		n := len(acc.AllowFilter) * len(current.AllowFilter)
+		allowFilter = make([]*qpN, 0, n)
+		for _, a := range acc.AllowFilter {
+			for _, c := range current.AllowFilter {
+				allowFilter = append(allowFilter, mkNodeFromLO(mkAndLogicalOperation([]*qpN{a, c})))
+			}
+		}
```

Comment:
For my own understanding, am I right in thinking that:

- if the previously evaluated scoped policy had "override parent" scoped permissions
- and therefore reaching this `mergePlans` function again meant that it didn't have a matching rule

That this cartesian product would yield no-op AND filters? E.g. predicates that are guaranteed to return false?

Reply:
No, this is a regular check with the parent policy.
Let's consider a few scenarios. In all of them, rules have the effect `ALLOW`.

Scenario 1: 
A child policy has a rule with the expression: `P.attr.x == 1` and the parent policy, whatever scopePermissions it has, has a rule with the expression: `P.attr.p == 'foo'`. Then, the result is a logical `AND` of these expressions.

Scenario 2:
A resource policy with default scopePermissions has two rules for this action with expressions: `P.attr.x == 1` and `P.attr.y == 3`. A resource with ether x equal one or y equal three will satisfy this policy, so the result is a logical `OR` of these expressions. This is why [here](https://github.com/cerbos/cerbos/blob/2fb58538db3e090847854975f5b90f0502844314/internal/engine/planner/planner.go#L154) we join them with `OR`.

Scenario 3 is a combination of these scenarios:
A child policy requiring parental consent has expressions A and B.
A parent policy with any scopePermission has expressions C and D. 
The result is:  `A && C || A && D || B && C || B && D`. Hence the cartesian product in `mergePlan`.

Whether the scopePermission is `OVERRIDE_PARENT` matters when I call `Complete` to check that we can exit the loop in the :
```go
	if !p.Empty() && p.ScopePermissions == policyv1.ScopePermissions_SCOPE_PERMISSIONS_OVERRIDE_PARENT {
		return true
	}
```
If the policy has a predicate and the `scopePermissions` equals to `OVERRIDE_PARENT`, then we can exit the loop.

I realised there's a bug. I call `acc.Complete()`, but I should call `currentResult.Complete()` instead. The `acc` won't be empty as it takes predicated from the child scopes.

Reply:
If the child policy had "override parent" and no matches, with expressions A and B, and
A parent policy has expressions C and D

Can't we say that A and B are guaranteed to be `false`, so all cartesian products `A && C || A && D || B && C || B && D` would always be `false` too?

Reply:
Sorry for the confusion. I meant that all expressions belong to the matched rules. Also, the expressions were intended to refer to unknown resource attributes, so I updated my comment to replace all `P.attr` occurrences with `R.attr`. 

> Can't we say that A and B are guaranteed to be false, so the cartesian product A && C || A && D || B && C || B && D would always be false too?

Taking my corrections into account, the answer is no. These expressions must appear in the Plan API response.